File: Orimath/Controls/IconImageConverter.cs
using System;
using System.Globalization;
using System.IO;
using System.Windows.Controls;
using System.Windows.Data;
using System.Windows.Media;
using System.Windows.Media.Imaging;

namespace Orimath.Controls
{
    [ValueConversion(typeof(Stream), typeof(Image))]
    public class IconImageConverter : IValueConverter
    {
        public object? Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            if (value is Stream stream)
            {
                using (stream)
                {
                    var source = new BitmapImage
                    {
                        CacheOption = BitmapCacheOption.OnLoad
                    };
                    source.BeginInit();
                    source.StreamSource = stream;
                    source.EndInit();

                    return new Image
                    {
                        Source = source,
                        Stretch = Stretch.Uniform,
                        Width = 16.0,
                        Height = 16.0,
                    };
                }
            }
            else
            {
                return null;
            }
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
        {
            throw new NotImplementedException();
        }
    }
}

File: Orimath/Controls/IconImageConverter.fs
namespace Orimath.Controls
open System
open System.IO
open System.Windows.Controls
open System.Windows.Data
open System.Windows.Media
open System.Windows.Media.Imaging

[<ValueConversion(typeof<Stream>, typeof<Image>)>]
type IconImageConverter() =
    interface IValueConverter with
        member _.Convert(value, _, _, _) =
            match value with
            | :? Stream as stream ->
                use stream = stream
                let source = BitmapImage()
                source.CacheOption <- BitmapCacheOption.OnLoad
                source.BeginInit()
                source.StreamSource <- stream
                source.EndInit()
                let result = AutoDisableImage()
                result.Source <- source
                result.Stretch <- Stretch.Uniform
                result.Width <- 16.0
                result.Height <- 16.0
                box result
            | _ -> null

        member _.ConvertBack(_, _, _, _) = raise (NotImplementedException())
